include/pqxx/connection.hxx

```diff
@@ -525,9 +615,49 @@ public:
    * there is incoming data on the connection's socket to be read, then call
    * @ref get_notifs repeatedly until it returns zero.
    *
-   * @return Number of notifications processed
+   * If your notifcation handler throws an exception, `get_notifs()` will just
+   * propagate it back to you.  (This is different from the old
+   * `notification_receiver` mechanism, which would merely log them.)
+   *
+   * @return Number of notifications processed.
    */
   int await_notification(std::time_t seconds, long microseconds);
+
+  /// A handler callback for notifications.
+  /** A notification handler takes 3 arguments: the _channel name_ to which
+   * the notification was sent; the process ID of the backend that sent the
+   * notification; and an optional "payload" text.  If there was no payload,
+   * you will receive an empty string there.
+   *
+   * Why the backend process ID?  Because sometimes you may want to send out
+   * a notification for every client listening on a channel, except yourself.
+   * If you sent out the notification yourself in the same session, then the
+   * handler will receive a process ID that's identical to the one that your
+   * connection's @ref backendpid() returns. 
+   */
+  using notification_handler = std::function<void(zview, int, zview)>;
```

Comment:
One thing I've been doing with this sort of API is to define a struct to pass in. The advantage here is that you can change the function signature without having to break clients:
```cpp
struct notification {
    zview name;
    int backend_pid;
    zview payload;
}
```

This gives you the ability to do a poor-man's named arguments when calling the notification callback, but it also means you can add arguments in the future without breaking backward compatibility.

Reply:
Nice idea, thanks!  I can't think of any extensions that might be coming, but who knows.  In any case it'll be easier for clients to ignore fields they don't care about, and (silly but I care) simple lambdas will be more compact and therefore easier to inline.

Reply:
Actually I _can_ think of an extension — many callbacks will want access to the connection.

Reply:
@KayEss I've implemented that change, and of course I'm pointlessly wondering whether there's any compelling to reason to pass the `notification` by reference (whether lvalue or rvalue).  Probably not.  I do like simplicity.

It does make the code a bit nicer.

Reply:
All compilers are fine with using the type name (I do this regularly), but you have to put the namespace name before the type, so something like `pqxx::connection connection` is fine. Here's a very similar example that works fine for clang, msvc and gcc (the `renderer` argument): <https://blue5alamander.com/open-source/planet-vk/include/planet/vk/engine/render_parameters.hpp>

Normally I would say that taking a `const &` makes the most sense, but I think that means the lambda would get a const reference to the connection, which is likely not good. it's worth remembering though that the lambda doesn't need to be an exact match for the `std::function` signature, it just needs to be callable from that signature. Here's an example: https://godbolt.org/z/s9c4r8xYd There are of course differences in the code generated in terms of when copies are created, but those are unlikely to matter at all in this context.

Reply:
Surprised to hear that `std::function` will take any compatibe signature.  In this case an rvalue reference would work well then, since it's a single-use object.  But... keeping things simple.